[BUG] withRestoreOnRetry should look at Throwable causes in addition to retry OOMs

Re: https://github.com/NVIDIA/spark-rapids/pull/7958

While debugging an issue with https://github.com/NVIDIA/spark-rapids/pull/7925 I ran into an issue where an UDF was failing to allocate a GPU buffer due to an injected OOM, but then it turns around and wraps it into its own exception:

```
Caused by: org.apache.spark.SparkException: Failed to execute user defined function: (emptygeneric: (string) => string)
```

This means the retry is hidden in the cause by, so I had to change the logic in the retry iterator to account for it. @jbrennan333, we'll need to adapt the new `withRestoreOnRetry` as well to account for this.

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/RmmRapidsRetryIterator.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/RmmRapidsRetryIterator.scala
@@ -230,13 +230,7 @@ object RmmRapidsRetryIterator extends Arm with Logging {
       case ex: Throwable =>
         // Only restore on retry exceptions
         val (topLevelIsRetry, _) = isRetryOrSplitAndRetry(ex)
-        val causedByRetry = topLevelIsRetry match {
-          case false =>
-            val (cbRetry, _) = causedByRetryOrSplit(ex)
-            cbRetry
-          case true => true
-        }
-        if (topLevelIsRetry || causedByRetry) {
+        if (topLevelIsRetry || causedByRetryOrSplit(ex)._1) {
           r.restore()
         }
         throw ex
@@ -263,13 +257,7 @@ object RmmRapidsRetryIterator extends Arm with Logging {
       case ex: Throwable =>
         // Only restore on retry exceptions
         val (topLevelIsRetry, _) = isRetryOrSplitAndRetry(ex)
-        val causedByRetry = topLevelIsRetry match {
-          case false =>
-            val (cbRetry, _) = causedByRetryOrSplit(ex)
-            cbRetry
-          case true => true
-        }
-        if (topLevelIsRetry || causedByRetry) {
+        if (topLevelIsRetry || causedByRetryOrSplit(ex)._1) {
           r.foreach(_.restore())
         }
         throw ex

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/RmmRapidsRetryIterator.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/RmmRapidsRetryIterator.scala
@@ -181,6 +181,36 @@ object RmmRapidsRetryIterator extends Arm with Logging {
   }
 
   /**
+   * Returns a tuple of (shouldRetry, shouldSplit) depending the exception
+   * passed
+   */
+  private def isRetryOrSplitAndRetry(ex: Throwable): (Boolean, Boolean) = {
+    ex match {
+      case _: RetryOOM => (true, false)
+      case _: SplitAndRetryOOM => (true, true)
+      case _ => (false, false)
+    }
+  }
+
+  /**
+   * Returns a tuple of (causedByRetry, causedBySplit) depending the exception
+   * passed
+   */
+  private def causedByRetryOrSplit(ex: Throwable): (Boolean, Boolean) = {
+    var current = ex
+    var causedByRetry = false
+    var causedBySplit = false
+    // check if there is a hidden retry or split OOM
+    while (current != null && !causedByRetry) {
+      current = current.getCause()
+      val (isRetry, isSplit) = isRetryOrSplitAndRetry(current)
+      causedByRetry = isRetry
+      causedBySplit = causedBySplit || isSplit
+    }
+    (causedByRetry, causedBySplit)
+  }
+
+  /**
    * withRestoreOnRetry for CheckpointRestore. This helper function calls `fn` with no input and
    * returns the result. In the event of an OOM Retry exception, it calls the restore() method
    * of the input and then throws the oom exception.  This is intended to be used within the `fn`
@@ -197,12 +227,19 @@ object RmmRapidsRetryIterator extends Arm with Logging {
     try {
       fn
     } catch {
-      case t: RetryOOM =>
-        r.restore()
-        throw t
-      case t: SplitAndRetryOOM =>
-        r.restore()
-        throw t
+      case ex: Throwable =>
+        // Only restore on retry exceptions
+        val (topLevelIsRetry, _) = isRetryOrSplitAndRetry(ex)
+        val causedByRetry = topLevelIsRetry match {
+          case false =>
+            val (cbRetry, _) = causedByRetryOrSplit(ex)
+            cbRetry
+          case true => true
+        }
+        if (topLevelIsRetry || causedByRetry) {
+          r.restore()
+        }
+        throw ex
     }
   }
 
@@ -223,12 +260,19 @@ object RmmRapidsRetryIterator extends Arm with Logging {
     try {
       fn
     } catch {
-      case t: RetryOOM =>
-        r.foreach(_.restore())
-        throw t
-      case t: SplitAndRetryOOM =>
-        r.foreach(_.restore())
-        throw t
+      case ex: Throwable =>
+        // Only restore on retry exceptions
+        val (topLevelIsRetry, _) = isRetryOrSplitAndRetry(ex)
+        val causedByRetry = topLevelIsRetry match {
+          case false =>
+            val (cbRetry, _) = causedByRetryOrSplit(ex)
+            cbRetry
+          case true => true
+        }
+        if (topLevelIsRetry || causedByRetry) {
+          r.foreach(_.restore())
+        }
+        throw ex
     }
   }
 
@@ -470,18 +514,6 @@ object RmmRapidsRetryIterator extends Arm with Logging {
       }
     }
 
-    /**
-     * Returns a tuple of (shouldRetry, shouldSplit) depending the exception
-     * passed
-     */
-    private def isRetryOrSplitAndRetry(ex: Throwable): (Boolean, Boolean) = {
-      ex match {
-        case retryOOM: RetryOOM => (true, false)
-        case splitAndRetryOOM: SplitAndRetryOOM => (true, true)
-        case _ => (false, false)
-      }
-    }
-
     override def next(): K = {
       // this is set on the first exception, and we add suppressed if there are others
       // during the retry attempts
@@ -513,20 +545,15 @@ object RmmRapidsRetryIterator extends Arm with Logging {
         } catch {
           case ex: Throwable =>
             // handle a retry as the top-level exception
-            val (topLevelIsRetry, isSplitAndRetry) = isRetryOrSplitAndRetry(ex)
-            doSplit = isSplitAndRetry
+            val (topLevelIsRetry, topLevelIsSplit) = isRetryOrSplitAndRetry(ex)
+            doSplit = topLevelIsSplit
 
             // handle any retries that are wrapped in a different top-level exception
             var causedByRetry = false
             if (!topLevelIsRetry) {
-              var current = ex
-              // check if there is a hidden retry OOM
-              while (current != null && !causedByRetry) {
-                current = current.getCause()
-                val (isRetry, isSplit) = isRetryOrSplitAndRetry(current)
-                causedByRetry = isRetry
-                doSplit = doSplit || isSplit
-              }
+              val (cbRetry, cbSplit) = causedByRetryOrSplit(ex)
+              causedByRetry = cbRetry
+              doSplit = doSplit || cbSplit
             }
 
             clearInjectedOOMIfNeeded()

--- a/tests/src/test/scala/com/nvidia/spark/rapids/WithRetrySuite.scala
+++ b/tests/src/test/scala/com/nvidia/spark/rapids/WithRetrySuite.scala
@@ -167,6 +167,31 @@ class WithRetrySuite
     }
   }
 
+  test("withRestoreOnRetry restores state on causedBy retry") {
+    val initialValue = 5
+    val increment = 5
+    var didThrow = false
+    val myCheckpointable = new SimpleCheckpointRestore(initialValue)
+    myCheckpointable.checkpoint()
+    try {
+      assertThrows[IllegalStateException] {
+        withRetryNoSplit {
+          withRestoreOnRetry(myCheckpointable) {
+            myCheckpointable.value += increment
+            if (!didThrow) {
+              val ex = new IllegalStateException()
+              ex.addSuppressed(new RetryOOM("causedby ex in tests"))
+              throw ex
+              didThrow = true
+            }
+          }
+        }
+      }
+    } finally {
+      assert(myCheckpointable.value == (initialValue + increment))
+    }
+  }
+
   test("withRestoreOnRetry restores state for Seq on retry") {
     val initialValue = 5
     val increment = 3
@@ -181,6 +206,33 @@ class WithRetrySuite
           if (!didThrow) {
             didThrow = true
             throw new RetryOOM("in tests")
+          }
+        }
+      }
+    } finally {
+      assert(myCheckpointables(0).value == (initialValue + increment))
+      assert(myCheckpointables(1).value == (initialValue + 10 + increment))
+    }
+  }
+
+  test("withRestoreOnRetry restores state for Seq on caused by retry") {
+    val initialValue = 5
+    val increment = 3
+    var didThrow = false
+    val myCheckpointables = Seq(new SimpleCheckpointRestore(initialValue),
+      new SimpleCheckpointRestore(initialValue + 10))
+    myCheckpointables.foreach(_.checkpoint())
+    try {
+      assertThrows[IllegalStateException] {
+        withRetryNoSplit {
+          withRestoreOnRetry(myCheckpointables) {
+            myCheckpointables.foreach(_.value += increment)
+            if (!didThrow) {
+              didThrow = true
+              val ex = new IllegalStateException()
+              ex.addSuppressed(new RetryOOM("causedby ex in tests"))
+              throw ex
+            }
           }
         }
       }

